### ISSUE TITLE
[feature request] provide a way to build without downloads

### ISSUE THREAD

**Issue by ghost** [asiekierka/MinecraftPOMF#2107]
First of all let me say I totally understand if you think this is out of scope for the project.
It would be really helpful for my work and mod(pack) development workflow though.

Basically I would like a mode to run all builds using the existing files in `~/.gradle/minecraft/`, assuming they are up-to-date and not attempting to connect to any servers to download them.

The goal is to 'package' mods to be able to build them from source and at some point even assemble a complete source-built modpack similar to what Linux distributions do with free software programs.

To cleanly build all projects and archive the dependencies, I need to seperate out the download step from all build processes. This means my packaging will take care of downloading the correct minecraft jars, libs and manifests, and the actual build process will run in a container without internet access.

These are my local changes to achive this, but they are uncoditional and break the normal download process:

```
diff --git a/build.gradle b/build.gradle
index f691ec36e..9f8f8f660 100644
--- a/build.gradle
+++ b/build.gradle
@@ -1,5 +1,6 @@
 buildscript {
 	repositories {
+		mavenLocal()
 		jcenter()
 		maven {
 			name "Fabric Repository"
@@ -10,13 +11,11 @@ buildscript {
 		classpath "cuchaz:enigma-cli:${project.enigma_version}"
 		classpath "net.fabricmc:stitch:${project.stitch_version}"
 		classpath "commons-io:commons-io:2.6"
-		classpath 'de.undercouch:gradle-download-task:4.0.4'
 		classpath 'net.fabricmc:tiny-remapper:0.3.1.72'
 	}
 }
 
 plugins {
-	id 'de.undercouch.download' version '4.0.4'
 	id 'base'
 	id 'maven-publish'
 	id 'java' // for jd gen
@@ -31,6 +30,7 @@ def build_number = ENV.BUILD_NUMBER ?: "local"
 def yarnVersion = "${minecraft_version}+build.$build_number"
 
 repositories {
+	mavenLocal()
 	mavenCentral()
 	maven {
 		name "Fabric Repository"
@@ -123,14 +123,8 @@ boolean validateChecksum(File file, String checksum) {
 
 task downloadVersionsManifest {
 	group = setupGroup
-	//inputs.property "mc_ver", minecraft_version
-	inputs.property "currenttime", new Date()
 	def manifestFile = new File(cacheFilesMinecraft, "version_manifest.json")
 	outputs.file(manifestFile)
-	doLast {
-		logger.lifecycle(":downloading minecraft versions manifest")
-		FileUtils.copyURLToFile(new URL("https://launchermeta.mojang.com/mc/game/version_manifest.json"), manifestFile)
-	}
 }
 
 def getManifestVersion(File manifestFile, String minecraft_version) {
@@ -142,81 +136,28 @@ def getManifestVersion(File manifestFile, String minecraft_version) {
 
 task downloadWantedVersionManifest(dependsOn: downloadVersionsManifest) {
 	group = setupGroup
-	def manifestFile = downloadVersionsManifest.outputs.files.singleFile
-	def manifestVersion = getManifestVersion(manifestFile, minecraft_version)
-
-	//have to grab the release time as there's a current timestamp on each element?!
-	inputs.property "releaseTime", manifestVersion.isPresent() ? manifestVersion.get().releaseTime : -1
-
 	outputs.file versionFile
 
-	doLast {
-		manifestVersion = getManifestVersion(manifestFile, minecraft_version)
-		//nb need to re-read here in case it didn't exist before
-		if (manifestVersion.isPresent() || versionFile.exists()) {
-
-			if (manifestVersion.isPresent()) {
-				FileUtils.copyURLToFile(new URL(manifestVersion.get().url), versionFile)
-			}
-		} else {
-			throw new RuntimeException("No version data for Minecraft version ${minecraft_version}")
-		}
-	}
-
 }
 
 task downloadMcJars(dependsOn: downloadWantedVersionManifest) {
 	group = setupGroup
-	inputs.files versionFile
-
 	outputs.files(clientJar, serverJar)
-
-	outputs.upToDateWhen {
-		def version = new JsonSlurper().parseText(FileUtils.readFileToString(versionFile))
-		return clientJar.exists() && serverJar.exists() && validateChecksum(clientJar, version.downloads.client.sha1) && validateChecksum(serverJar, version.downloads.server.sha1)
-	}
-
-	doLast {
-		if (!versionFile.exists()) {
-			throw new RuntimeException("Can't download the jars without the ${versionFile.name} file!")
-		}
-
-		//reload in case it changed
-		def version = new JsonSlurper().parseText(FileUtils.readFileToString(versionFile))
-
-		logger.lifecycle(":downloading minecraft jars")
-
-		download {
-			src new URL(version.downloads.client.url)
-			dest clientJar
-			overwrite false
-		}
-
-		download {
-			src new URL(version.downloads.server.url)
-			dest serverJar
-			overwrite false
-		}
-	}
 }
 
-task downloadIntermediary(type: Download) {
+task downloadIntermediary() {
 	group = buildMappingGroup
-	def url = "https://github.com/FabricMC/intermediary/raw/master/mappings/${minecraft_version}.tiny"
-	src com.google.common.net.UrlEscapers.urlFragmentEscaper().escape(url)
-	dest new File(cacheFilesMinecraft, "${minecraft_version}-intermediary.tiny")
+	outputs.file(new File(cacheFilesMinecraft, "${minecraft_version}-intermediary.tiny"))
 }
 
-task downloadIntermediaryV2(type: Download) {
+task downloadIntermediaryV2() {
 	group = buildMappingGroup
-	def url = "https://maven.fabricmc.net/net/fabricmc/intermediary/${minecraft_version}/intermediary-${minecraft_version}-v2.jar"
-	src com.google.common.net.UrlEscapers.urlFragmentEscaper().escape(url)
-	dest new File(cacheFilesMinecraft, "${minecraft_version}-intermediary-v2.jar")
+	outputs.file(new File(cacheFilesMinecraft, "${minecraft_version}-intermediary-v2.jar"))
 }
 
 task extractIntermediaryV2(dependsOn: downloadIntermediaryV2, type: Copy) {
 	def output = new File(cacheFilesMinecraft, "${minecraft_version}-intermediary-v2.tiny")
-	from({ zipTree(downloadIntermediaryV2.dest) }) {
+	from({ zipTree(downloadIntermediaryV2.outputs.files.singleFile) }) {
 		include 'mappings/mappings.tiny'
 		rename 'mappings.tiny', "../${output.name}"
 	}
@@ -247,42 +188,12 @@ task mergeJars(dependsOn: downloadMcJars) {
 
 task downloadMcLibs(dependsOn: downloadWantedVersionManifest) {
 	group = setupGroup
-	inputs.files versionFile
-
 	outputs.dir(libraries)
-
-	outputs.upToDateWhen { false }
-
-	doLast {
-		if (!versionFile.exists()) {
-			throw new RuntimeException("Can't download the jars without the ${versionFile.name} file!")
-		}
-
-		def version = new JsonSlurper().parseText(FileUtils.readFileToString(versionFile, StandardCharsets.UTF_8))
-
-		logger.lifecycle(":downloading minecraft libraries")
-
-		if (!libraries.exists()) {
-			libraries.mkdirs()
-		}
-
-		version.libraries.each {
-			def downloadUrl = it.downloads.artifact.url
-
-			download {
-				src downloadUrl
-				dest new File(libraries, downloadUrl.substring(downloadUrl.lastIndexOf("/") + 1))
-				overwrite false
-			}
-
-			project.dependencies.add("decompileClasspath", it.name)
-		}
-	}
 }
 
 task invertIntermediary(dependsOn: downloadIntermediary, type: FileOutput) {
 	group = buildMappingGroup
-	def v1Input = downloadIntermediary.dest
+	def v1Input = downloadIntermediary.outputs.files.singleFile
 
 	output = new File(cacheFilesMinecraft, "${minecraft_version}-intermediary-inverted.tiny")
 	outputs.file(output)
@@ -356,7 +267,7 @@ task mapIntermediaryJar(dependsOn: [downloadMcLibs, downloadIntermediary, mergeJ
 
 	doLast {
 		logger.lifecycle(":mapping minecraft to intermediary")
-		def tinyInput = downloadIntermediary.dest
+		def tinyInput = downloadIntermediary.outputs.files.singleFile
 		mapJar(intermediaryJar, mergedFile, tinyInput, libraries, "official", "intermediary")
 	}
 }
@@ -517,7 +428,7 @@ clean.doFirst {
 tasks.build.dependsOn "compressTiny","tinyJar","v2UnmergedYarnJar", "v2MergedYarnJar"
 
 task exportMappingsOfficial(dependsOn: downloadIntermediary) {
-	def composeInput = downloadIntermediary.dest
+	def composeInput = downloadIntermediary.outputs.files.singleFile
 
 	doLast {
 		logger.lifecycle(":exporting mappings")

**Comment by modmuss50** [comment]
Im not sure yarn is what you want to be looking at, https://github.com/FabricMC/fabric-loom is what is used to build mods, you can run it with --offline but this doesnt sound like it would acomplish what you are after.

I do have plans (and some stuff locally) to start making offline installer bundles, basicaly a zip containing everything required to **run** mods without using stuff hosted on our infra(intermediary, loader and libs). This does not include stuff provided by mojang.

You do bring up an intresting point about building mods without the support of fabric's servers, the installer bundle I have planned would sitll be useful but doesnt cover the case of loom + it's deps.

To fullfill the goal I think you have will require you to attack the loom/gradle side of things, as yarn can easily be archived from the maven: https://maven.fabricmc.net/net/fabricmc/yarn/.

**Comment by ghost** [comment]
> Im not sure yarn is what you want to be looking

Kinda true. yarn is a dependency to run loom, so I'm looking at it first, because I want to build all minecraft-related projects used in the entire build from source. Specifically I need the generated tiny files.
But you are right, I will have the same problem when running any gradle build with the fabric-loom plugin. It won't simply work with `--offline`, even when all maven dependencies are mirrored into a  locale maven mirror. I'll open the corresponding issues in fabric-loom when I know the details on what would need to be changed ;)

**Comment by modmuss50** [comment]
I dont think there is any work to do on our end for this, but im happy to try and point you in the right direction. For yarn I would just change over the urls where it pulls from. (Setup your own maven with the required deps). The loom side should be similar if you want to replace just our stuff. 

The easyist way that I can think of to achieve this would be to setup a caching poxy server.

**Comment by liach** [comment]
Hmm, are you editing minecraft mappings offline? If you don't need to edit mappings, you can just use the existing distributions.

If you do need to edit minecraft mappings offline, your patch looks fine. However, I suggest you try minimize your patches if you want to keep your patch maintainable.

**Comment by liach** [comment]
Also, @petabyteboy I wonder what your use case is (if you can disclose it). I don't think this way of building yarn can solve many problems, and I might find better solutions for you.

**Comment by ghost** [comment]
TL;DR: There are many use cases for this. Editing minecraft mappings offline is not a use case that's important to me, but archival of the modding infrastructure for later use or integration of all the components are.

It's more of a philosophical question. yarn as well as most fabric mods out there are free software. It's very much possible for a user to modify them, build them from source and publish modified changes. However, using them to build a software distribution (modpack) is really hard, because you would need to rebuild a lot of small components from source. This is something a source-based package manager can automate, similar to how you can build a whole operationg system from source with Gentoo's portage, Exherbo's Paludis, NixOS' Nix, GuixSD's Guix or Yocto's OpenEmbedded build system.

A practical use case: I built a modpack using free software Fabric mods only. One motivation behind this were crashes because some mods were partially built against different fix release of Minecraft, or there were built against different releases of each other. For example many of the mods are using loom 0.4 and built against Minecraft 1.16.3. Now 1.16.5 is released, I update the Minecraft version of the modpack, and now I would like to know which of the twenty mods might need changes to work well with the new Minecraft versions.

From my experience, having such packaging for software means people will do awesome things with it, because it's much easier to experiment with changes. You press one button, and it will rebuild an entire modpack from source with the change you made in some base component of the toolchain.

There's another story to be told here. Recently I wanted to play a Minecraft 1.2.5 modpack with a few friends for the good old nostalgia. I tried to find a zip with the server components, but I could only find some self-extracting compressed executable, that would try to download a forge jar from some server that ceised existing in the last nine years. It was a pain to find all binary assets. A friend tried, but in addition to the server being down, their computer isn't using the x86_64 instruction set, so they couldn't even run this packed server installer executable.
Having a source-based packaging for all components will prevent a situation like this. As long as the source repositories are still around or I have archived them (which I can easily do with my package manager as well if I have build recipes btw), I can build the mods. On any architecture and without any dependency on all the Fabric maven servers still being around.
There are projects for archival of the software many of use everyday with a similar approach: https://www.tweag.io/blog/2020-06-18-software-heritage/

**Comment by liach** [comment]
for your purpose, i suggest you to just download one yarn mapping from aven and stick to that mapping for each minecraft version for your pre-offline setup. loom supports passing local mapping jars as mapping dependencies.

**Comment by ghost** [comment]
The same could be said of any other library or piece of software/data required for the build process of a mod. I could think of many use cases specifically for building a local modified version of yarn from source and using it in the build process of another mod, without manually copying around jar files.
However since this is probably not of much use to anyone but me, I'll just maintain my own patches on top of the yarn (and loom) gradle files.
Thanks for looking at the problem!
